sys/aarch64/gpio.c

```diff
@@ -0,0 +1,46 @@
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <aarch64/bcm2835reg.h>
+#include <aarch64/bcm2835_gpioreg.h>
+
+/*
+ * \brief delay function
+ * \param delay number of cycles to delay
+ *
+ * This just loops <delay> times in a way that the compiler
+ * won't optimize away.
+ */
+static void delay(int64_t count) {
+  __asm__ volatile("1: subs %[count], %[count], #1; bne 1b"
+                   : [count] "+r"(count));
+}
+
+void gpio_function_select(resource_t *r, unsigned pin, unsigned func) {
+  unsigned mask = (1 << BCM2835_GPIO_GPFSEL_BITS_PER_PIN) - 1;
+  unsigned reg = pin / BCM2835_GPIO_GPFSEL_PINS_PER_REGISTER;
+  unsigned shift = (pin % BCM2835_GPIO_GPFSEL_PINS_PER_REGISTER) *
+                   BCM2835_GPIO_GPFSEL_BITS_PER_PIN;
+
+  unsigned val = bus_read_4(r, BCM2835_GPIO_GPFSEL(reg));
+  val &= ~(mask << shift);
+  val |= (func << shift);
+  bus_write_4(r, BCM2835_GPIO_GPFSEL(reg), val);
+}
+
+void gpio_set_pull(resource_t *r, unsigned pin, unsigned pud) {
+  assert(pin <= 40);
```

Comment:
Magic constant alert. Are pins counted from 0 or 1 ?

Reply:
It depends...
Now should be better.